spirv_msl.cpp

```diff
@@ -2572,6 +2575,26 @@ void CompilerMSL::add_composite_member_variable_to_interface_block(StorageClass
 			return;
 	}
 
+	// Recursively handle nested structures.
+	if (mbr_type.basetype == SPIRType::Struct)
+	{
+		for (uint32_t i = 0; i < elem_cnt; i++)
+		{
+			string mbr_name = append_member_name(mbr_name_qual, var_type, mbr_idx) + (elem_cnt == 1 ? "" : join("_", i));
+			string var_chain = join(var_chain_qual, ".", to_member_name(var_type, mbr_idx), (elem_cnt == 1 ? "" : join("[", i, "]")));
+			uint32_t sub_mbr_cnt = uint32_t(mbr_type.member_types.size());
+			for (uint32_t sub_mbr_idx = 0; sub_mbr_idx < sub_mbr_cnt; sub_mbr_idx++)
+			{
+				add_composite_member_variable_to_interface_block(storage, ib_var_ref, ib_type,
+																 var, mbr_type, sub_mbr_idx,
+																 meta, mbr_name, var_chain,
+																 location, var_mbr_idx);
+				var_mbr_idx++;
```

Comment:
With recursion, cannot add_composite_member_variable_to_interface_block increment more than once? Maybe var_mbr_idx should be mutable and incremented by the add_composite_member_variable function itself?

Reply:
Yes. Recursive incrementation happens here, because at this point we are inside `add_composite_member_variable_to_interface_block()`, and calling it for the next level.

Could move the incrementing into `add_composite_member_variable_to_interface_block()`, but would also have to also move it into `add_plain_member_variable_to_interface_block()`, and still handle the masking that doesn't invoke either.

Reply:
I still don't understand the purpose this exact line. I tried a test shader like this:

```
#version 450
layout(quads) in;

layout(location = 0) out vec4 FragColor;

struct Bar
{
	float c;
	float d;
};

struct Foo
{
	Bar a;
	Bar b;
};

layout(location = 0) in Foo foos[];

void main()
{
	FragColor.x = foos[0].a.c;
	FragColor.y = foos[1].a.d;
	FragColor.z = foos[0].b.c;
	FragColor.w = foos[1].b.d;
}
```
and this ends up emitting code like:
```
#include <metal_stdlib>
#include <simd/simd.h>

using namespace metal;

struct Bar
{
    float c;
    float d;
};

struct Foo
{
    Bar a;
    Bar b;
};

struct main0_out
{
    float4 FragColor [[user(locn0)]];
    float4 gl_Position [[position]];
};

struct main0_in
{
    float Foo_a_c [[attribute(0)]];
    float Foo_a_d [[attribute(1)]];
    float Foo_b_c [[attribute(2)]];
    float Foo_b_d [[attribute(3)]];
};

struct main0_patchIn
{
    patch_control_point<main0_in> gl_in;
};

[[ patch(quad, 0) ]] vertex main0_out main0(main0_patchIn patchIn [[stage_in]])
{
    main0_out out = {};
    out.FragColor.x = patchIn.gl_in[0].Foo_a_c;
    out.FragColor.y = patchIn.gl_in[1].Foo_a_d;
    out.FragColor.z = patchIn.gl_in[0].Foo_a_d; <-- Misaligned?
    out.FragColor.w = patchIn.gl_in[1].Foo_b_c;
    return out;
}
```
If I understand it correctly, var_mbr_idx is supposed to uniquely define a member inside the composite object, but something seems off, e.g.:

```
		set_extended_member_decoration(ib_type.self, ib_mbr_idx, SPIRVCrossDecorationInterfaceMemberIndex, var_mbr_idx);
```
Here var_mbr_idx takes 0, 1, 3, 4. 2 is skipped due to this line incrementing things additionally.

Reply:
Hmmm...after digging into this more...the problem is not really how `var_mbr_idx` is calculated. Moving it inside `add_composite_member_variable_to_interface_block()` and `add_plain_member_variable_to_interface_block()` doesn't help this use case, and it in turn breaks `shaders-msl/asm/frag/pull-model-interpolation.asm.msl23.frag` and other test shaders.

The problem for this use case is that `emit_tessellation_access_chain()`, `emit_tessellation_io_load()`, `fix_up_interpolant_access_chain()`, and `fix_up_interface_member_indices()` all need to handle the recursion as well.

There is so much existing complex and repetitive code surrounding this, particularly around tessellation handling, that this is going to get really ugly. I did some work on this earlier, but pulled back to the current PR state because of that. 

But to do this right, covering all recursive structures is going to take some significant rework. That's all too much to take care of right now. I'm going to move this PR to WIP for now. I'll try to find some time to properly address this soon.

Reply:
Sure, as long as the code is sane and doesn't cause regressions we can merge.